### PR TITLE
JIT: enable removal of try/catch if the try can't throw.

### LINKED ISSUE #107191: JIT does not elide empty try/catch blocks
### Description

If we want to format some runtime values for a `Debug.WriteLine` invocation, it's possible for the formatting method to throw. So, I figured I'd wrap a try/catch around the `Debug.WriteLine` call. However, the JIT does not elide the block in release mode.

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgBEBLbAcwDsJcMGxcBuAWABQZSrQBKAV1Zd8MGgGEI+AA4MANjCgBlTQDduMPkJHUaAFRgIMAwcYDMlAEzl5avLiEBvIeV/kA2gCyMBgAFhAAJgCSKmoAFMFhkTHKagDyylwQrLg0AILMzLC4uAy6MBkyDABe2FmsAJQAuj5+xA7EKOSJ4RFxwACeGDD+TeQRddgNrb7egn4L5BhQAzOLc4ub5HQwwBLMNADqUAzDADIMrDBxAKKskBGXBwCqZgBiABw0AOIhWstPOITDBTBo2LbkAC+awWYDqYFCML8Gwhfh2ewOx1OMAuVziACJgNgImoBuQAGbQfB1YYRcaTfFgpG+aHzPysyFAA===

<details><summary>C# / IL / JIT ASM</summary>

```csharp
using System.Diagnostics;
using System.Runtime.CompilerServices;
using System.Text;

public class Class
{
    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    public void Method(byte[] data)
    {
        try
        {
            Debug.WriteLine(Encoding.UTF8.GetString(data));
        }
        catch
        {
            Debug.WriteLine("badly formatted data");
        }
    }
}
```

```asm
.method public hidebysig 
    instance void Method (
        uint8[] data
    ) cil managed flag(0200) 
{
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableContextAttribute::.ctor(uint8) = (
        01 00 01 00 00
    )
    // Method begins at RVA 0x2050
    // Code size 6 (0x6)
    .maxstack 1

    .try
    {
        IL_0000: leave.s IL_0005
    } // end .try
    catch [System.Runtime]System.Object
    {
        IL_0002: pop
        IL_0003: leave.s IL_0005
    } // end handler

    IL_0005: ret
} // end of method Class::Method
```

```asm
; Core CLR 8.0.724.31311 on x64

Class..ctor()
    L0000: ret

Class.Method(Byte[])
    L0000: push rbp
    L0001: sub rsp, 0x10
    L0005: lea rbp, [rsp+0x10]
    L000a: mov [rbp-0x10], rsp
    L000e: add rsp, 0x10
    L0012: pop rbp
    L0013: ret
    L0014: push rbp
    L0015: sub rsp, 0x10
    L0019: mov rbp, [rcx]
    L001c: mov [rsp], rbp
    L0020: lea rbp, [rbp+0x10]
    L0024: lea rax, [L000e]
    L002b: add rsp, 0x10
    L002f: pop rbp
    L0030: ret
```

</details>

### Configuration

8.0.724.31311 on x64

### Analysis

The assembly appears to set up a stack frame, immediately tear it down, then return. Curiously, there are unreachable instructions after the `ret` on `L0013`.

Alternatively, the JIT *will* elide an empty try/*finally* block, reducing `Class.Method` to a simple `ret`.

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgBEBLbAcwDsJcMGxcBuAWABQZSrQAqMBBgGChxAMyUATOQDCAGzy4hAbyHkDlRcRTkAsjAwALCABMAFMACeGGAG0AuuVvYM2AJT6hnqChmHkGFBOQeHkIbGxdDDAAK7MNADqUAyuADIMrDD2AKKskLYF6QCqYgBiABw0AOKWAMqRlfY+fv7+MgnkAL4x4QBmBdjq6tGhCfEDhkmp6Vk5MPmF9gBEW30jYcOzBoeDQA

<details><summary>C# / IL / JIT ASM</summary>

```csharp
using System.Diagnostics;
using System.Text;

public class Class
{
    public void Method(byte[] data)
    {
        try
        {
            Debug.WriteLine(Encoding.UTF8.GetString(data));
        }
        finally
        {
            Debug.WriteLine("");
        }
    }
}
```

```asm
.method public hidebysig 
    instance void Method (
        uint8[] data
    ) cil managed 
{
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableContextAttribute::.ctor(uint8) = (
        01 00 01 00 00
    )
    // Method begins at RVA 0x2050
    // Code size 4 (0x4)
    .maxstack 0

    .try
    {
        IL_0000: leave.s IL_0003
    } // end .try
    finally
    {
        IL_0002: endfinally
    } // end handler

    IL_0003: ret
} // end of method Class::Method
```

```asm
; Core CLR 8.0.724.31311 on x64

Class..ctor()
    L0000: ret

Class.Method(Byte[])
    L0000: ret
```

</details>

### PR DESCRIPTION
**Pull request by AndyAyersMS**
If no tree in the try region of a try/catch can throw, then we can remove the try and delete the catch.

If no tree in the try region of a try/finally can throw, we can remove the try and inline the finally. This slightly generalizes the empty-try/finally opt we have been doing for a long time. (We should do something similar for try/fault, but don't, yet).

Since these optimization passes are cheap, and opportunities for them arise after other optimizations and unblock subsequent optimizations, run them early, middle, and late.

Resolves #107191.

I expect we'll see more of these cases in the future, say if we unblock cloning of loops with EH.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@EgorBo PTAL
cc @dotnet/jit-contrib 

Note we are trusting `GTF_EXCEPT` fairly early... I think this is ok (it is glob ref that is iffy). Should be small number of diffs, but potential for more down the road with some work I have in flight.

File: src/coreclr/jit/fgehopt.cpp [hunk 1]
```diff
@@ -659,7 +759,298 @@ PhaseStatus Compiler::fgRemoveEmptyTry()
...
+
+        if (hadDfs)
+        {
+            fgInvalidateDfsTree();
+            m_dfsTree = fgComputeDfs();
+            if (hadLoops)
+            {
+                m_loops = FlowGraphNaturalLoops::Find(m_dfsTree);
+            }
+        }
```
Comment (jakobbotsch) [inline src/coreclr/jit/fgehopt.cpp:1053]: It seems unnecessary to recompute this in each of the phases. Maybe combine it into one phase, or make the actual phase that needs these annotations compute them if they aren't computed when we get there?

**Comment by VSadov** [comment]
> If no tree in the try region of a try/finally can throw, we can remove the try and inline the finally. 

Just wondering - We used to have some special meaning for code in finally with respect to ThreadAbort. Is that no longer a thing or still somehow maintained when finally is inlined?

**Comment by MichalPetryka** [comment]
> > If no tree in the try region of a try/finally can throw, we can remove the try and inline the finally.
> 
> Just wondering - We used to have some special meaning for code in finally with respect to ThreadAbort. Is that no longer a thing or still somehow maintained when finally is inlined?

AFAIR all the handling for that was removed.

**Comment by AndyAyersMS** [comment]
> > > If no tree in the try region of a try/finally can throw, we can remove the try and inline the finally.
> > 
> > 
> > Just wondering - We used to have some special meaning for code in finally with respect to ThreadAbort. Is that no longer a thing or still somehow maintained when finally is inlined?
> 
> AFAIR all the handling for that was removed.

That is only an issue for .NET Framework. Removal of empty-try / finally has been in .NET Core since 2.x, this just extends it a bit further to some trys that are not empty.

**Comment by VSadov** [comment]
Replace search with bing on ios


Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: src/coreclr/jit/fgehopt.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/fgehopt.cpp]: moved recomputation later (to doms for dfs, vn for loops).

**Review by jakobbotsch — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
> But the codegen of `Test3` which has both `catch` and `finally` is still not optimal:

This turns into a try-catch inside a try-finally. We remove the try-catch and convert the try-finally to a try-fault. But we haven't yet implemented (empty)try-fault removal.

**Comment by AndyAyersMS** [comment]
Seems like this is actually simple to handle; I'll put up a PR shortly.

**Pull request merged**